generals/envs/generalsio_client.py

```diff
@@ -0,0 +1,109 @@
+from socketio import SimpleClient  # type: ignore
+
+from generals.agents.agent import Agent
+
+
+class GeneralsBotError(Exception):
+    """Base generals-bot exception
+    TODO: find a place for exceptions
+    """
+
+    pass
+
+
+class GeneralsIOClientError(GeneralsBotError):
+    """Base GeneralsIOClient exception"""
+
+    pass
+
+
+class RegisterAgentError(GeneralsIOClientError):
+    """Registering bot error"""
+
+    def __init__(self, msg: str) -> None:
+        super().__init__()
+        self.msg = msg
+
+    def __str__(self) -> str:
+        return f"Failed to register the agent. Error: {self.msg}"
+
+
+class GeneralsIOClient(SimpleClient):
+    """
+    Wrapper around socket.io client to enable Agent to join
+    GeneralsIO lobby.
+    """
+
+    def __init__(self, agent: Agent, user_id: str):
+        super().__init__()
+        self.connect("https://botws.generals.io")
+        self.user_id = user_id
+        self._queue_id = ""
+
+    @property
+    def queue_id(self):
+        if not self._queue_id:
+            raise GeneralsIOClientError("Join the game before waiting for it to start.")
```

Comment:
what does this error mean?

Reply:
Basically that agent is waiting for game before joining the game. I realise now that the exception message is a bit unfortunate in a way that it is not exception message but more of a suggestion message. I will reword it to something more suitable and put suggestion at the end.